pipeline/lib/pipeline.ml

```diff
@@ -74,79 +76,109 @@ let github_status_of_state url = function
+module Lifecycle = struct
+  let build ~repo_id ~pull_number ~branch ~dockerfile ~head db =
+    let src =
+      match head with
+      | `Github api_commit ->
+          Git.fetch
+            (Current.map Github.Api.Commit.id (Current.return api_commit))
+      | `Local commit -> commit
+    in
+    let* commit =
+      match head with
+      | `Github api_commit -> Current.return (Github.Api.Commit.hash api_commit)
+      | `Local commit -> commit >>| Git.Commit.hash
+    in
+    let current_image = Docker.build ~pool ~pull:false ~dockerfile (`Git src) in
+    let* build_job_id = Current_util.get_job_id current_image in
+    (* FIXME: compute in db *)
+    let run_at = Ptime_clock.now () in
+    match build_job_id with
+    | Some build_job_id ->
+        Storage.record_build_start ~run_at ~repo_id ~pull_number ~branch ~commit
+          ~build_job_id db;
+        Current.map (fun image -> (build_job_id, image)) current_image
+    | None -> failwith "No build job id"
+
+  let run ~tmpfs ~docker_cpuset_cpus ~docker_cpuset_mems ~repo_id ~head db
+      image_build =
     let run_args =
       [ "--security-opt"; "seccomp=./aslr_seccomp.json" ]
       @ tmpfs
       @ docker_cpuset_cpus
       @ docker_cpuset_mems
     in
-    let run_at = Ptime_clock.now () in
     let* commit =
       match head with
       | `Github api_commit -> Current.return (Github.Api.Commit.hash api_commit)
       | `Local commit -> commit >>| Git.Commit.hash
     in
-    let repo_info = owner ^ "/" ^ repository in
+    let* build_job_id, _ = image_build in
+    let current_image = Current.map snd image_build in
     let current_output =
-      Docker_util.pread_log ~run_args current_image ~repo_info ?pull_number
-        ?branch ~commit
+      Docker_util.pread_log ~run_args current_image ~repo_info:"TODO" ~commit
         ~args:
           [
             "/usr/bin/setarch"; "x86_64"; "--addr-no-randomize"; "make"; "bench";
           ]
     in
-    let+ build_job_id = Current_util.get_job_id current_image
-    and+ run_job_id = Current_util.get_job_id current_output
-    and+ output = current_output in
-    let duration = Ptime.diff (Ptime_clock.now ()) run_at in
-    Logs.debug (fun log -> log "Benchmark output:\n%s" output);
-    let () =
-      let db = new Postgresql.connection ~conninfo () in
-      output
-      |> Json_util.parse_many
-      |> List.iter (fun output_json ->
-             let benchmark_name =
-               Yojson.Safe.Util.(member "name" output_json)
-               |> Yojson.Safe.Util.to_string_option
-             in
-             Yojson.Safe.Util.(member "results" output_json)
-             |> Yojson.Safe.Util.to_list
-             |> List.map
-                  (Benchmark.make ~duration ~run_at ~repo_id ~benchmark_name
-                     ~commit ?pull_number ?build_job_id ?run_job_id ?branch)
-             |> List.iter (Models.Benchmark.Db.insert db));
-      db#finish
+    let* run_job_id = Current_util.get_job_id current_output
```

Comment:
will `run_job_id` only be determined once docker has run with the make bench target? Would we still have a `pending stage`? I think the only thing left to do here is to put the pread_logs in the database.